File: Leetcode Contest/Biweekly Contest 174/Number of Alternating XOR Partitions.py
from typing import List
from collections import defaultdict

class Solution:
    MOD = 1_000_000_007

    def alternatingXOR(self, nums: List[int], target1: int, target2: int) -> int:
        n = len(nums)

        # dp1[i]: ways where the last block XOR = target1
        # dp2[i]: ways where the last block XOR = target2
        dp1 = [0] * (n + 1)
        dp2 = [0] * (n + 1)

        map1 = defaultdict(int)  # prefix_xor -> sum of dp1
        map2 = defaultdict(int)  # prefix_xor -> sum of dp2

        px = 0

        for i in range(1, n + 1):
            px ^= nums[i - 1]

            # block XOR = target1
            ways1 = map2[px ^ target1]
            if px == target1:
                ways1 += 1  # single block from start
            dp1[i] = ways1 % self.MOD

            # block XOR = target2 (must follow target1)
            dp2[i] = map1[px ^ target2] % self.MOD

            map1[px] = (map1[px] + dp1[i]) % self.MOD
            map2[px] = (map2[px] + dp2[i]) % self.MOD

        return (dp1[n] + dp2[n]) % self.MOD

File: Leetcode Contest/Biweekly Contest 174/Number of Alternating XOR Partitions.java
import java.util.*;

class Solution {
    static final int MOD = 1_000_000_007;

    public int alternatingXOR(int[] nums, int target1, int target2) {
        int n = nums.length;

        long[] dp1 = new long[n + 1]; // last block = target1
        long[] dp2 = new long[n + 1]; // last block = target2

        Map<Integer, Long> map1 = new HashMap<>();
        Map<Integer, Long> map2 = new HashMap<>();

        int px = 0;

        for (int i = 1; i <= n; i++) {
            px ^= nums[i - 1];

            // block XOR = target1
            long ways1 = map2.getOrDefault(px ^ target1, 0L);
            if (px == target1) ways1++; // single-block from start
            dp1[i] = ways1 % MOD;

            // block XOR = target2 (must come after target1)
            dp2[i] = map1.getOrDefault(px ^ target2, 0L) % MOD;

            map1.put(px, (map1.getOrDefault(px, 0L) + dp1[i]) % MOD);
            map2.put(px, (map2.getOrDefault(px, 0L) + dp2[i]) % MOD);
        }

        // valid partitions must end with either target1 or target2
        return (int)((dp1[n] + dp2[n]) % MOD);
    }
}
